src/relay/pass/inline.cc

```diff
@@ -0,0 +1,227 @@
+    auto base_func = call_graph_->GetGlobalFunction(global);
+    const auto* fn = base_func.as<FunctionNode>();
+    CHECK(fn) << "Expected to work on a Relay function.";
+
+    auto func = FunctionNode::make(fn->params,
+                                   fn->body,
+                                   fn->ret_type,
+                                   fn->type_params,
+                                   fn->attrs);
+    // Inline the function body to the caller if this function uses default
+    // compiler, i.e. no external codegen is needed.
+    if (func->UseDefaultCompiler()) {
+      CHECK_EQ(func->params.size(), args.size())
+          << "Mismatch found in the number of parameters and call args";
+      // Bind the parameters with call args.
+      Map<Var, Expr> bind_map;
+      for (size_t i = 0; i < args.size(); i++) {
+        bind_map.Set(fn->params[i], args[i]);
+      }
+      if (const auto* gvn = expr.as<GlobalVarNode>()) {
+        auto ret_type = gvn->checked_type();
+        // Cannot replace TensorType/TensorTupleType with FuncType. Therefore,
+        // we simply inline the function as a closure instead of directly using
+        // its body when the global var returns FuncType.
+        return ret_type->IsInstance<FuncTypeNode>() ? std::move(func)
+                                                    : func->body;
+      } else {
+        return Bind(func->body, bind_map);
+      }
+    } else if (const auto* call_node = expr.as<CallNode>()) {
+        return CallNode::make(func, args, call_node->attrs, call_node->type_args);
+    } else {
+      return std::move(func);
+    }
+  }
+
+  /*!
+   * \brief The current call graph entry that is being handled. Each entry
+   * contains a global function.
+   */
+  CallGraphEntry* cur_node_;
+  /*! \brief The call graph that is used for global function lookup. */
+  const CallGraphNode* call_graph_;
+};
+
+IRModule Inline(const IRModule& module) {
+  CallGraph cg(module);
+  auto topo = cg->TopologicalOrder();
+  // Get the reverse topological order of the global functions.
+  std::reverse(topo.begin(), topo.end());
+  // Cache the functions that are originally entries. These functions will
+  // remain in the module after inlining.
+  std::unordered_set<CallGraphEntry*> original_entry;
+
+  for (auto* it : topo) {
+    if (it->GetRefCount() == 0) original_entry.emplace(it);
+    // Skip the leaf calls and the recursive calls that don't call other
+    // functions.
+    if (it->empty() || (it->IsRecursive() && it->size() == 1)) continue;
+    auto base_func = module->Lookup(it->GetNameHint());
+    if (const auto* fn = base_func.as<FunctionNode>()) {
+      auto func = GetRef<Function>(fn);
+      auto new_func = Inliner(it, cg.operator->()).Inline(func);
+      // TODO(zhiics) Maybe move this to CallGraph, but updating function from
+      // CallGraph arbitarily may lead to incorrect CallGraph.
+      cg->module->Update(it->GetGlobalVar(), new_func);
```

Comment:
Just for my own curiosity 
* Why are you updating the call graph? Can call graph be a side data structure? Like infer type, whenever a Relay expr is transformed, the side data structure needs to be re-generated.
* If the call graph is modified, is the earlier collected topological order still ok?

Reply:
1. Updating the call graph here essentially updates the Relay IRModule because the functions are maintained there. Yes, we can always re-generate the call graph for different passes, but    it would be good to just maintain one of it in the same pass since constructing a call graph is also costly.

2. It is okay since we are also maintaining the edges here, but this should be very careful.

Reply:
For the first point, what is a preferred option in general? Should we always update an analyzed side-data-structures whenever a mutator pass runs? Do you know if LLVM udates everytime/never updates/updates for some and not for other side data structures. In general, it will be good to set this as a design principle for analysis passes.

For this PR, it is ok.

Reply:
LLVM's pass manager is much more powerful. It globs a bunch of passes as preserved. For optimizations that mutate the IR, they will have some checkers/validators to check which analysis results are still valid. The analyses be still preserved if they are still valid.